Add cylindrical selection test using PeriodicKDTree

rishinamdev17 opened the pull request:
Fixes #1736

## Changes made in this Pull Request:
- Added a test for cylindrical distance-based atom selections (`cylayer`) using `PeriodicKDTree`
- Ensured the KDTree-based distance search produces the same results as the selection engine
- Covers periodic boundary conditions for cylindrical selections

## PR Checklist
- [x] Issue raised/referenced?
- [x] Tests updated/added?
- [ ] Documentation updated/added? (not required for test-only change)
- [ ] `package/CHANGELOG` file updated? (not required for test-only change)
- [ ] Is your name in `package/AUTHORS`? (optional for small contributions)

## Developers Certificate of Origin

I certify that I can submit this code contribution as described in the  
[Developer Certificate of Origin](https://developercertificate.org/),  
under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).

rishinamdev17 committed:
Add cylindrical selection test using PeriodicKDTree

BradyAJohnston commented:
Hi @rishinamdev17 - we can't review this PR at all given the tests are currently failing. Please look into ensuring the tests are passing (ideally you wouldn't open a new PR for a new test that doesn't pass to begin with).

rishinamdev17 committed:
Format cylindrical selection test with pinned Black version

rishinamdev17 commented:
hi @BradyAJohnston 
Thanks for your patience. I wanted to let you know that the cylindrical selection test I added is still failing locally due to a parsing issue (TypeError: float() argument must be a string or a real number, not 'NoneType'). I’m actively investigating this and will update the PR once the test passes.

Appreciate any guidance or suggestions if there’s a recommended way to handle this selection.

Thanks!
Rishii

rishinamdev17 commented:
hi @BradyAJohnston 
Thanks for the feedback — I’ve been debugging this further and the failure is coming from the selection parser itself rather than the KDTree logic.

The cylayer selection appears to internally route through PointSelection and expects additional tokens (cutoff) that aren’t obvious from the public API, which results in a float(None) error even with valid-looking syntax like cylayer … index 0.

I don’t want to force a workaround without understanding the intended grammar, so I’m pausing further changes until I get clarification on the correct cylayer selection form to use for this test. Happy to update once advised.

BradyAJohnston commented:
You are not currently using `cylayer` properly as it expects a selection to use as the COG. Please read the [documentation](https://userguide.mdanalysis.org/stable/selections.html): 

> selects all atoms within a cylindric layer centered in the center of geometry (COG) of a given selection, e.g. `cylayer 5 10 10 -8 protein` selects the center of geometry of protein, and creates a cylindrical layer of inner radius 5, external radius 10 centered on the COG. In z, the cylinder extends from 10 above the COG to 8 below. Positive values for zMin, or negative ones for zMax, are allowed.

rishinamdev17 committed:
Add sanity test for cylayer selection using COG

rishinamdev17 committed:
Format cylayer selection test for Black

rishinamdev17 commented:
Thanks for your patience @BradyAJohnston !!
I’m still tracking down the remaining Black formatting failure locally. The cylayer selection logic now looks correct following your guidance, and I’m aligning my local Black setup with the project configuration so the final CI check passes. I’ll push an update shortly.

tylerjereddy commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -1731,9 +1695,7 @@ def test_chirality(smi, chirality):
-    assert_equal(
-        u.atoms[:3].chiralities, np.array(["", chirality, ""], dtype="U1")
-    )
+    assert_equal(u.atoms[:3].chiralities, np.array(["", chirality, ""], dtype="U1"))
```
Are all these unrelated whitespace changes really needed?

rishinamdev17 replied:
Those whitespace changes are unrelated to my test. I'll revert them and keep the PR focused only on the cylayer test.

tylerjereddy commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -33,6 +33,7 @@
 from MDAnalysis import SelectionError, SelectionWarning
 from MDAnalysis.core.selection import Parser
 from MDAnalysis.lib.distances import distance_array
+from MDAnalysis.lib.pkdtree import PeriodicKDTree
```
Where is this being used?

rishinamdev17 replied:
Yes you are right!! that import is no lomger needed after simplifying the test and I have removed it..

tylerjereddy commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -1774,3 +1736,20 @@ def test_formal_charge_selection(sel, size, name):
+def universe():
+    u = mda.Universe(PSF, DCD)
+    u.dimensions = np.array([100.0, 100.0, 100.0, 90.0, 90.0, 90.0])
+    return u
```
Creating a fixture for use in a single test case seems a bit much, especially when you're repeating the setting of `dimensions` below anyway.

rishinamdev17 replied:
I have combined the universe setup into the test and removed the duplicate dimensions setting.

rishinamdev17 committed:
Address review feedback for cylayer selection test

BradyAJohnston commented:
There are still a lot of formatting changes that should be undone. The test you have also now created no longer really tests what you set out to do - it is just testing the parsing of a selection and not that it matches against using the PeriodicKDTree.

orbeckst commented:
@BradyAJohnston thanks for moving the PR along. I assigned it to you, given that you're already doing the work of the PR shepherd. Thank you!

Aryaman-Chaudhri commented:
Hi @rishinamdev17 , I noticed you are struggling with black , you might want to look into CONTRIBUTING.md.

rishinamdev17 committed:
Test cylayer selection against PeriodicKDTree reference

rishinamdev17 commented:
Hi @BradyAJohnston @tylerjereddy, I’ve reverted the unrelated formatting changes and updated the test so it validates cylayer behavior against an explicit geometric reference. All checks are now passing... Ready for your review!!

rishinamdev17 commented:
Hey! @BradyAJohnston @tylerjereddy ... just a gentle ping in case this slipped through. All checks are still passing and I’m happy to make any changes if needed!!

BradyAJohnston commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -1774,3 +1776,32 @@ def test_formal_charge_selection(sel, size, name):
+    dxy = np.sqrt((pos[:, 0] - center[0]) ** 2 + (pos[:, 1] - center[1]) ** 2)
+    dz = pos[:, 2] - center[2]
+
+    mask = (inner <= dxy) & (dxy <= outer) & (zmin <= dz) & (dz <= zmax)
```
The `CylindricalSelection._apply` actually uses strict inequalities (`<` and `>` rather than `<=` and `>=` so the test should be changed to reflect this.

https://github.com/MDAnalysis/mdanalysis/blob/c32fe45f9d6083fe3d1fb681bc8e1fdbb94aa6c9/package/MDAnalysis/core/selection.py#L516

BradyAJohnston commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -1774,3 +1776,32 @@ def test_formal_charge_selection(sel, size, name):
+
+def test_cylayer_selection_matches_periodic_kdtree():
+    u = mda.Universe(PSF, DCD)
+    u.dimensions = np.array([100.0, 100.0, 100.0, 90.0, 90.0, 90.0])
```
These dimensions are too large to actually test the PBC for cylindrical selections. You would ned to use a different trajectory or change the dimensions.
